File: CYC.py
# reference:
# H.M.Edwards "Fermat's Last Theorem" section 4.2

import numpy as np
import sympy as sp
from math import gcd

class CYC:
    """ cyclotomic integers """
    @staticmethod
    def init(n):
        """ set degree n of cyclotomic integer
        assume n is odd prime
        """
        if n<3 or not sp.isprime(n):
            raise RuntimeError('wrong n')
        n1,g = n-1, sp.primitive_root(n)
        CYC.n, CYC.n1 = n,n1
        CYC.g_pow = np.empty(n1, dtype=np.int)
        CYC.log_g = np.empty(n, dtype=np.int)
        CYC.cjind = np.empty((n1,n), dtype=np.int)
        CYC.order = np.empty(n, dtype=np.int)
        a = 1; CYC.order[0] = 1
        for i in range(n1):
            CYC.g_pow[i] = a
            CYC.log_g[a] = i
            CYC.order[a] = n1//gcd(i,n1)
            CYC.cjind[-i] = np.r_[:a*n:a] % n
            a = a * g % n

    @staticmethod
    def ord(a): # order of a mod n; ord(0)==1
        return CYC.order[a%CYC.n]

    def __init__(a, c=0):
        """ set coefficients c of cyclotomic integer
           as a polynomial of degree n-1.
        c: 1d-array or scalar of int
           in order of increasing power of omega
           where omega = exp(2*pi*i/n)
        """
        a.c = np.asarray(c, dtype='object')
        a.c.resize(a.n)

    def __hash__(a): return hash(sum(a.c[::2]))
    def __repr__(a): return str(np.trim_zeros(a.c, 'b'))
    def __neg__(a): return CYC(-a.c) # -a

    # make degree of a less than n-1
    def normalize(a): return CYC(a.c - a.c[-1])

    # conjugate by replacing omega with omega^{g^i}
    def conj(a, i=1): return CYC(a.c[a.cjind[i]])

    # product of all conjugates of a
    def norm(a):
        b = [a.conj(i) for i in range(a.n1)]
        return np.prod(b, axis=0).c[0]

    # return content and primitive part of a
    def primitive(a):
        d = 0
        for c in a.c: d = gcd(c,d)
        return d, CYC(a.c//d)

    def __add__(a,b): # a+b
        if isinstance(b,CYC): return CYC(a.c + b.c)
        else: return CYC(a.c + b)

    def __sub__(a,b): # a-b
        if isinstance(b,CYC): return CYC(a.c - b.c)
        else: return CYC(a.c - b)

    def __mul__(a,b): # a*b
        # b is either CYC or int
        if not isinstance(b,CYC):
            return CYC(b * a.c)
        c = np.convolve(a.c, b.c)
        c[:a.n1] += c[a.n:]
        return CYC(c[:a.n] - c[a.n1]) # normalize

    def __truediv__(a,b):
        """ test if b divides a.
        return None if b doesn't divide a,
        return b/a if b divides a.
        """
        if isinstance(b,CYC):
            c = [b.conj(i) for i in range(1,b.n1)]
            c = np.prod(c, axis=0)
            a *= c; b = (b*c).c[0]
        if np.any(a.c%b): return None
        else: return CYC(a.c//b)

    def to_int(a): # const term afeter normalizing
        return a.c[0] - a.c[-1]

    def isRational(a):
        """ test if a is rational
        return None if a is not rational
        return a.to_int() if a is raitonal
        """
        if np.all(a.c[1:] == a.c[1]):
            return a.to_int()
        else: return None

    def __eq__(a,b): # equality testing
        # b is either CYC or int
        if isinstance(b,CYC):
            return np.all(a.c - a.c[-1] ==
                          b.c - b.c[-1])
        else:
            a = a.isRational()
            if not a: return False
            else: return a == b

    def __pow__(a,e):
        """ a^e, assume e>=0 """
        if e==0: return CYC(1)
        m = (1<<(e.bit_length()-1))>>1
        b = a
        while m:
            b *= b
            if e&m: b *= a
            m>>=1
        return b

    def isUnit(a): # test if a is unit
        return a.norm() == 1

    def isAssoc(a,b):
        """ test if a and b are associates """
        q = a/b
        if q: return q.isUnit()
        else: return False

    def __radd__(a,b): return CYC(b + a.c)
    def __rsub__(a,b): return CYC(b - a.c)
    def __rmul__(a,b): return CYC(b * a.c)

    @staticmethod
    def random(b=1):
        """ random cyclotomic integer
        with |coefficients| < b
        """
        return CYC(np.random.randint(-b+1, b, CYC.n1))

File: C++/CYC.cpp
// reference:
// H.M.Edwards "Fermat's Last Theorem" section 4.2

// uses NTL
//   http://www.shoup.net/ntl

#include "CYC.h"
#include "ZZFactoring.h"
#include<exception>
using namespace NTL;

long CYC::n; // degree + 1
long CYC::g; // primitive root mod n
ZZX CYC::modulus; // cyclotomic polynomial
Vec<long> CYC::g_pow; // g^j % n (j=0...n-1)
Vec<long> CYC::log_g; // ind_g a (a=1...n-1)
Vec<long> CYC::order; // order of a mod n (a=1...n-1)
Mat<long> CYC::cjind; // j*g^i%n (i=0...n-1,j=0...n)

long primitive_root(long);

void CYC::init(long n_)
// set degree n of cyclotomic integer
// assume n is odd prime
{
    if(n_<3 || !ProbPrime(n_))
        throw std::runtime_error("wrong n in CYC::init");
    n = n_;
    long i,j, n1(n-1), a(1);
    for(i=0; i<n; i++)
        SetCoeff(modulus, i);

    g = primitive_root(n);
    g_pow.SetLength(n1);
    log_g.SetLength(n);
    cjind.SetDims(n1,n);
    order.SetLength(n); order[0] = 1;
    for(i=0; i<n1; i++) {
        g_pow[i] = a;// g^j mod n
        log_g[a] = i;// ind_g a
        order[a] = n1/GCD(i,n1);// order of a mod n
        cjind[i][0] = 0;
        for(j=1; j<n; j++)// indices of conjugates
            cjind[i][j] = AddMod(cjind[i][j-1], a, n);
        a = MulMod(a,g,n);
    }
}

long IsUnit(const CYC& a)
// test if norm(a)==1
{ ZZ n; norm(n,a); return IsOne(n); }

long IsAssoc(const CYC& a, const CYC& b)
// test if a/b is unit
{ CYC q; return divide(q,a,b) && IsUnit(q); }

void SetCoeff(CYC& a, long i, const ZZ& c)
{// a[i] = c
    SetCoeff((ZZX&)a, i%=a.n, c);
    if(i==a.n-1) a %= a.modulus;
}

void SetCoeff(CYC& a, long i, long c)
{// a[i] = c
    SetCoeff((ZZX&)a, i%=a.n, c);
    if(i==a.n-1) a %= a.modulus;
}

void SetCoeff(CYC& a, long i)
{// a[i] = 1
    SetCoeff((ZZX&)a, i%=a.n);
    if(i==a.n-1) a %= a.modulus;
}

void conv(CYC& a, const NTL::vec_ZZ& c)
{// a[i] = c[i] (i=0...n-1)
    conv((ZZX&)a, c);
    a %= a.modulus;
}

CYC& operator*=(CYC& b, const CYC& a)// b *= a
{ MulMod(b, b, a, b.modulus); return b; }

void conj(CYC& b, const CYC& a, long i)
// b = i-th conjugate of a
// replacing omega by omega^{g^i}
{
    if(&b==&a) { CYC c(a); conj(b,c); return; }
    long j;
    b.SetLength(a.n);
    for(j=0; j<=deg(a); j++)
        b[b.cjind[i][j]] = a[j];
    for(j=deg(a)+1; j<a.n; j++)
        clear(b[b.cjind[i][j]]);
    b %= b.modulus;
}

long divide(const CYC& a, const CYC& b)
// test if b divides a and
// n = norm(b) is optionally supplied.
{
    long i;
    CYC c,d,s(b),t(1);
    for(i=2; i<a.n; i++) { conj(s,s); t *= s; }
    mul(c,a,t);
    mul(d,b,t);
    for(i=0; i<=deg(c); i++)
        if(!divide(c[i], d[0])) return 0;
    return 1;
}

long divide(CYC& q, const CYC& a, const CYC& b)
// test if b divides a and
// set q = a/b if b divides a, else q is unchanged.
// n = norm(b) is optionally supplied.
{
    long i;
    ZZ u[q.n];
    CYC c,d,s(b),t(1);
    for(i=2; i<q.n; i++) { conj(s,s); t *= s; }
    mul(c,a,t);
    mul(d,b,t);
    for(i=0; i<=deg(c); i++)
        if(!divide(u[i], c[i], d[0])) return 0;
    clear(q);
    for(i=0; i<=deg(c); i++) SetCoeff(q, i, u[i]);
    return 1;
}

void power(CYC& b, const CYC& a, long n)
// b = a^n; assume n>=0
{
    if(n==0) { set(b); return; }
    if(&b==&a) { CYC c(a); power(b,c,n); return; }
    long m(1<<(NumBits(n)-1));
    b=a;
    for(m>>=1; m; m>>=1) {
        sqr(b,b);
        if(n&m) b*=a;
    }
}

void RandomBnd(CYC& x, const ZZ& n)
// x = random cyclotomic integer
//     with |coefficients| less than n
{
    x.SetLength(x.n-1);
    for(long i=x.n-2; i>=0; i--) {
        RandomBnd(x[i], n);
        if(RandomBits_long(1))
            negate(x[i], x[i]);
    }
    x.normalize();
}